src/lib/support/CHIPMemString.cpp

```diff
@@ -31,17 +31,21 @@ namespace Platform {
 
 char * MemoryAllocString(const char * string, size_t length)
 {
-    if (length == 0)
+    if (string && !length)
     {
         length = strlen(string);
```

Comment:
This can overrun the end. Are there many uses of `length=0` in calls? Strongly suggest avoiding this particular usage of len==0, unless there is a sane max length provided.

Reply:
I find the contract odd.  Is this supposed to be something like a `strndup()`?

Reply:
Only one caller actually supplies a length — all the rest were previously `strdup()` —and would just change from `MemoryAlloc(s)` to `MemoryAlloc(s, strlen(s))`. The allocated size is always `length+1`.

Reply:
> I find the contract odd. Is this supposed to be something like a `strndup()`?

Yes, this replaced str[n]dup() calls.

Reply:
this does not replicate `strndup()`...

```
#include <stdio.h>
#include <string.h>

int main()
{
    char * p = strndup("hi", 0);
    printf("p = %p\n", p);
    if (p)
    {
        printf("strlen(p) = %lu\n", strlen(p));
    }
}
```
yields:
```
gcc -o strndup strndup.c && ./strndup
p = 0x7fd8e6c058f0
strlen(p) = 0


```

Reply:
other aspects of the strndup() contract not met with this change:
* `strndup(null, 0)` returns the empty string
* `strndup(null, 1)` aborts

Reply:
I'll replace the default parameter with an overload to match the 0 case. Passing NULL to str* is UB, though, and I'd rather provide defined behaviour — the current choice is to treat a null source as an empty string; would you prefer an abort()?

Reply:
blind dereference of the pointer would be my preference.  extra behavior is more to document, harder to debug.

Reply:
My main concern is the fact that this may cause possibly very large allocations on passing-in of garbage or carefully crafted data. I feel like `strdup()` is always incorrect in embedded, and in fact, I have worked on many coding guidelines that forbade the "n"-less string functions (e.g. strlen, strdup, strcpy) and mandated strnlen, strndup, strncpy/strlcpy, etc. I feel like this "convenience" string duplicate with no upper bounds provided by caller may not be desirable on constrained devices.

Reply:
much larger change to keep folks from calling strlen() on the interrupt vector... ;)

Reply:
OK. I guess it'll make it a little easier on someone later just to hunt down all the `strlen`s.